tests/src/test/scala/akka/kafka/scaladsl/RebalanceExtSpec.scala

```diff
@@ -0,0 +1,410 @@
+        )
+      // consumer-2::SubSource-topic-1-1-1-A:confirm first messageId=10 is received and committed from batch (10,11,12)
+      topicMetadata.messageStore.msgAckMap(10).waitUntil.complete(Try(Done))
+      Await.result(topicMetadata.messageStore.msgAckMap(10).ackWaitUntil.future, remainingOrDefault)
+      // consumer-2::SubSource-topic-1-1-1-A:verify messageId=10 is received in the business logic function
+      assert(topicMetadata.messageStore.msgAckMap(10).messageCounter.intValue() == 1)
+
+      // consumer-2::define post-abort partition distribution
+      AlpakkaAssignor.clientIdToPartitionMap.set(
+        Map(
+          consumerClientId2 -> Set(t1p0, t1p1)
+        )
+      )
+      // consumer-1::SubSource-topic-1-1-0-A:unblock messageId=2 from batch (1,2,3)
+      topicMetadata.messageStore.msgAckMap(2).waitUntil.complete(Try(Done))
+      // consumer-1::SubSource-topic-1-1-0-A:verify messageId=2 thread is unblocked
+      Await.result(topicMetadata.messageStore.msgAckMap(2).ackWaitUntil.future, remainingOrDefault)
+      // assert(topicMetadata.messageStore.msgAckMap(2).messageCounter.intValue() == 1)
+
+      // consumer-1::SubSource-topic-1-1-0-A:abort at messageId=2 (uncommitted)
+      sharedKillSwitch1.abort(new Throwable(s"abort $consumerClientId1 messageId=2"))
+
+      // consumer-2::after abort two new sub sources serve topic-1-1-0 and topic-1-1-1: SubSource-topic-1-1-0-B and SubSource-topic-1-1-1-B
+
+      // consumer-2::SubSource-topic-1-1-1-A:unblock second message from batch (10,11,12)
+      topicMetadata.messageStore.msgAckMap(11).waitUntil.complete(Try(Done))
+      Await.result(topicMetadata.messageStore.msgAckMap(11).ackWaitUntil.future, remainingOrDefault)
+      // assert(topicMetadata.messageStore.msgAckMap(11).messageCounter.intValue() == 1)
+
+      // consumer-2::SubSource-topic-1-1-0-B starts consuming at its first batch (2,3,4)
+      // consumer-2::SubSource-topic-1-1-1-B starts consuming at its first batch (12,13,14)
+
+      // consumer-1::SubSource-topic-1-1-0-A:unblock last message from batch (1,2,3)
+      // consumer-2::SubSource-topic-1-1-0-B:unblock second message from batch (2,3,4)
+      topicMetadata.messageStore.msgAckMap(3).waitUntil.complete(Try(Done))
+      Await.result(topicMetadata.messageStore.msgAckMap(3).ackWaitUntil.future, remainingOrDefault)
+
+      // consumer-1::SubSource-topic-1-1-0-A:Terminates
+
+      // consumer-2::SubSource-topic-1-1-0-B:unblock last message from batch (2,3,4)
+      topicMetadata.messageStore.msgAckMap(4).waitUntil.complete(Try(Done))
+      Await.result(topicMetadata.messageStore.msgAckMap(4).ackWaitUntil.future, remainingOrDefault)
+      // assert(topicMetadata.messageStore.msgAckMap(2).messageCounter.intValue() == 2) // message replay
+      // assert(topicMetadata.messageStore.msgAckMap(3).messageCounter.intValue() == 2) // message replay
+      // assert(topicMetadata.messageStore.msgAckMap(4).messageCounter.intValue() == 1)
```

Comment:
Necessary?

Reply:
They were added to verify the notes and code in the associated section.